pkg/controller/certificatesigningrequests/vault/vault.go

```diff
@@ -0,0 +1,159 @@
+	cmapi "github.com/jetstack/cert-manager/pkg/apis/certmanager/v1"
+	experimentalapi "github.com/jetstack/cert-manager/pkg/apis/experimental/v1alpha1"
+	controllerpkg "github.com/jetstack/cert-manager/pkg/controller"
+	"github.com/jetstack/cert-manager/pkg/controller/certificatesigningrequests"
+	"github.com/jetstack/cert-manager/pkg/controller/certificatesigningrequests/util"
+	internalvault "github.com/jetstack/cert-manager/pkg/internal/vault"
+	logf "github.com/jetstack/cert-manager/pkg/logs"
+	"github.com/jetstack/cert-manager/pkg/util/pki"
+)
+
+const (
+	CSRControllerName = "certificatesigningrequests-issuer-vault"
+)
+
+type signingFn func(*x509.Certificate, *x509.Certificate, crypto.PublicKey, interface{}) ([]byte, *x509.Certificate, error)
+
+// Vault is a controller for signing Kubernetes CertificateSigningRequest
+// using Vault Issuers.
+type Vault struct {
+	issuerOptions controllerpkg.IssuerOptions
+	secretsLister corelisters.SecretLister
+
+	recorder record.EventRecorder
+
+	certClient    certificatesclient.CertificateSigningRequestInterface
+	clientBuilder internalvault.ClientBuilder
+}
+
+func init() {
+	controllerpkg.Register(CSRControllerName, func(ctx *controllerpkg.Context) (controllerpkg.Interface, error) {
+		return controllerpkg.NewBuilder(ctx, CSRControllerName).
+			For(certificatesigningrequests.New(apiutil.IssuerVault, NewVault(ctx))).
+			Complete()
+	})
+}
+
+func NewVault(ctx *controllerpkg.Context) *Vault {
+	return &Vault{
+		issuerOptions: ctx.IssuerOptions,
+		secretsLister: ctx.KubeSharedInformerFactory.Core().V1().Secrets().Lister(),
+		recorder:      ctx.Recorder,
+		certClient:    ctx.Client.CertificatesV1().CertificateSigningRequests(),
+		clientBuilder: internalvault.New,
+	}
+}
+
+// Sign attempts to sign the given CertificateSigningRequest based on the
+// provided Vault Issuer or ClusterIssuer. This function updates the
+// CertificateSigningRequest resource if signing was successful. Returns an
+// error which, if not nil, should trigger a retry.
+func (v *Vault) Sign(ctx context.Context, csr *certificatesv1.CertificateSigningRequest, issuerObj cmapi.GenericIssuer) error {
+	log := logf.FromContext(ctx, "sign")
+	log = logf.WithRelatedResource(log, issuerObj)
+
+	resourceNamespace := v.issuerOptions.ResourceNamespace(issuerObj)
+
+	client, err := v.clientBuilder(resourceNamespace, v.secretsLister, issuerObj)
+	if apierrors.IsNotFound(err) {
+		message := "Required secret resource not found"
+		log.Error(err, message)
+		v.recorder.Event(csr, corev1.EventTypeWarning, "SecretNotFound", message)
+		util.CertificateSigningRequestSetFailed(csr, "SecretNotFound", message)
+		_, err := v.certClient.UpdateStatus(ctx, csr, metav1.UpdateOptions{})
+		return err
+	}
+
+	if err != nil {
+		message := fmt.Sprintf("Failed to initialise vault client for signing: %s", err)
+		log.Error(err, message)
+		v.recorder.Event(csr, corev1.EventTypeWarning, "ErrorVaultInit", message)
+		return err
```

Comment:
suggestion: `return nil` since this is a non-retryable error
```suggestion
        // No need to retry since this is a permanent error (until the user changes
        // the Secret or the Issuer).
		return nil
```

Reply:
The `New` function here is doing a few things which has error handling throughout, including making API calls to vault. These network calls should be retried.

Reply:
Would it be possible to add a comment on top of vault's `New` that indicates that errors are transient/are retryable?

Reply:
Done.

https://github.com/jetstack/cert-manager/pull/4103/files#diff-5f677db86560eb3dbd3b363d179ecf99987739859455e3e33a5edc49cbf6e1dbR73